File: Medium/Javascript/452. Minimum Number of Arrows to Burst Balloons.js
/**Given the array points, return the minimum number of arrows that must be shot to burst all balloons. */

/**
 * @param {number[][]} points
 * @return {number}
 */
var findMinArrowShots = function (points) {
  points.sort((a, b) => a[0] - b[0]);

  let stack = [points[0]];
  let topOfStack = points[0];

  for (let i = 1; i < points.length; i++) {
    if (points[i][0] <= topOfStack[1]) {
      stack.pop();
      let element = [
        Math.max(points[i][0], topOfStack[0]),
        Math.min(points[i][1], topOfStack[1]),
      ];
      stack.push(element);
      topOfStack = element;
    } else {
      stack.push(points[i]);
      topOfStack = points[i];
    }
  }

  return stack.length;
};

console.log(
  findMinArrowShots([
    [1, 2],
    [3, 4],
    [5, 6],
    [7, 8],
  ])
);

File: Medium/python/452. Minimum Number of Arrows to Burst Balloons.py
'''
There are some spherical balloons taped onto a flat wall that represents the XY-plane. The balloons are represented as a 2D integer array points where points[i] = [xstart, xend] denotes a balloon whose horizontal diameter stretches between xstart and xend. You do not know the exact y-coordinates of the balloons.

Arrows can be shot up directly vertically (in the positive y-direction) from different points along the x-axis. A balloon with xstart and xend is burst by an arrow shot at x if xstart <= x <= xend. There is no limit to the number of arrows that can be shot. A shot arrow keeps traveling up infinitely, bursting any balloons in its path.

Given the array points, return the minimum number of arrows that must be shot to burst all balloons.
'''


class Solution:
    def findMinArrowShots(self, points: list[list[int]]) -> int:
        points.sort(key=lambda x: x[0])
        stack = [points[0]]

        for location in points:
            if location[0] <= stack[-1][1]:
                element = stack.pop()
                stack.append([max(element[0], location[0]), min(element[1], location[1])])
            else:
                stack.append(location)

        return len(stack)

print(Solution().findMinArrowShots([[10,16],[2,8],[1,6],[7,12],[3,9]]))